File: examples/hello-tinygo/http.go
package main

import (
	"encoding/json"
	"errors"
	"strconv"
	"strings"
)

// HTTP egress through the host (docs/abi.md, "HTTP egress"): the guest never
// opens a socket; it hands the host a request and the host performs it within
// the Composition's sandbox.egress grant and the operator's egress policy, or
// answers with a refusal. The wire format is JSON both ways; the host writes
// its response into a buffer allocated through this guest's own wasmfn_alloc.

// HTTPRequest is one request for the host to perform.
type HTTPRequest struct {
	// Method of the request; empty means GET.
	Method string `json:"method,omitempty"`
	// URL, http or https, absolute.
	URL string `json:"url"`
	// Headers to send. Host, Content-Length and hop-by-hop headers are the
	// host's to set and are dropped.
	Headers map[string][]string `json:"headers,omitempty"`
	// Body bytes (base64 on the wire, as encoding/json renders []byte).
	Body []byte `json:"body,omitempty"`
}

// HTTPResponse is what the server answered: its status, whatever it is (a
// 503 is a response, not an error), headers and body.
type HTTPResponse struct {
	Status  int                 `json:"status"`
	Headers map[string][]string `json:"headers,omitempty"`
	Body    []byte              `json:"body,omitempty"`
	// Error is set instead of a status when the host did not perform the
	// request: refused by the grant or the policy, over a budget, or failed.
	Error string `json:"error,omitempty"`
}

// HTTPError is the host's reason for not performing a request. It is a
// distinct type so a guest can tell a refusal from a server error.
type HTTPError struct{ Reason string }

func (e *HTTPError) Error() string { return e.Reason }

// httpSink hands a JSON request to the host and returns its JSON response;
// the wasip1 build points it at the wasmfn.http import, other builds have no
// host and refuse — a native test substitutes its own.
var httpSink = noHostHTTP

func noHostHTTP([]byte) ([]byte, error) {
	return nil, errors.New("wasmfn: no host HTTP in this build (not running under function-wasm)")
}

// HTTPGet performs a GET through the host.
func HTTPGet(url string) (*HTTPResponse, error) {
	return HTTPDo(&HTTPRequest{Method: "GET", URL: url})
}

// HTTPDo performs req through the host. A request the host does not perform
// is an *HTTPError naming the reason; a response from the server, whatever
// its status, is returned as is.
func HTTPDo(req *HTTPRequest) (*HTTPResponse, error) {
	payload, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	out, err := httpSink(payload)
	if err != nil {
		return nil, err
	}
	rsp := &HTTPResponse{}
	if err := json.Unmarshal(out, rsp); err != nil {
		return nil, errors.New("wasmfn: cannot decode the host's HTTP response: " + err.Error())
	}
	if rsp.Error != "" {
		return nil, &HTTPError{Reason: rsp.Error}
	}
	return rsp, nil
}

// httpGetText performs a GET and returns the body of a 200 as text, trimmed;
// any other status is an error naming it.
func httpGetText(url string) (string, error) {
	rsp, err := HTTPGet(url)
	if err != nil {
		return "", err
	}
	if rsp.Status != 200 {
		return "", errors.New("GET " + url + ": status " + strconv.Itoa(rsp.Status))
	}
	return strings.TrimSpace(string(rsp.Body)), nil
}

File: examples/hello-assemblyscript/assembly/http.ts
// HTTP egress through the host (docs/abi.md, "HTTP egress"): the guest never
// opens a socket; it hands the host a JSON request and the host performs it
// within the egress grant of the module's manifest and the operator's policy,
// or answers with a refusal. The response body is base64 both ways.

import { JSON } from "json-as";

import { hostHttp } from "./host";


@json
class WireRequest {
  method: string = "GET";
  url: string = "";
}


@json
class WireResponse {
  status: i32 = 0;
  body: string | null = null;
  error: string | null = null;
}

// The outcome of a request: text is the trimmed body of a 200, err the host's
// reason for anything else. AssemblyScript has no exceptions, so errors travel
// by value.
export class HttpText {
  text: string = "";
  err: string | null = null;
}

// getText GETs url through the host and returns the whitespace-trimmed body
// of a 200; any other answer (a refusal, a non-200 status, a payload this
// guest cannot read) is an err.
export function getText(url: string): HttpText {
  const out = new HttpText();
  const request = new WireRequest();
  request.url = url;
  const payload = String.UTF8.encode(JSON.stringify(request));
  const packed = hostHttp(changetype<usize>(payload), <u32>payload.byteLength);
  if (packed == 0) {
    out.err = "the host returned no response";
    return out;
  }
  const raw = String.UTF8.decodeUnsafe(<usize>(packed >>> 32), <usize>(packed & 0xffffffff));
  const response = JSON.parse<WireResponse>(raw);
  if (response.status == 0) {
    const reason = response.error;
    out.err = reason !== null && reason.length > 0 ? reason : "the host returned no status and no error";
    return out;
  }
  if (response.status != 200) {
    out.err = "GET " + url + ": status " + response.status.toString();
    return out;
  }
  const body = response.body;
  if (body === null) {
    return out;
  }
  const decoded = decodeBase64(body);
  if (decoded === null) {
    out.err = "the host's HTTP response body is not base64";
    return out;
  }
  out.text = decoded.trim();
  return out;
}

const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

// decodeBase64 decodes the standard padded alphabet (how the host renders
// body bytes) into a string, or null when s is not base64.
function decodeBase64(s: string): string | null {
  const bytes = new Array<u8>();
  let acc: u32 = 0;
  let bits: i32 = 0;
  for (let i = 0; i < s.length; i++) {
    const c = s.charAt(i);
    if (c == "=") break;
    const v = alphabet.indexOf(c);
    if (v < 0) return null;
    acc = (acc << 6) | <u32>v;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      bytes.push(<u8>((acc >> bits) & 0xff));
    }
  }
  const buf = new ArrayBuffer(bytes.length);
  for (let i = 0; i < bytes.length; i++) {
    store<u8>(changetype<usize>(buf) + i, bytes[i]);
  }
  return String.UTF8.decode(buf);
}
